config/store.go

```diff
@@ -188,124 +196,155 @@ func (s *Store) Set(newCfg *model.Config) (*model.Config, error) {
 		return nil, errors.Wrap(err, "new configuration is invalid")
 	}
 
-	newCfg = removeEnvOverrides(newCfg, s.configNoEnv, s.GetEnvironmentOverrides())
+	newCfgNoEnv := removeEnvOverrides(newCfg, oldCfgNoEnv, s.GetEnvironmentOverrides())
 
-	// Don't persist feature flags unless we are on MM cloud
+	// Don't store feature flags unless we are on MM cloud
 	// MM cloud uses config in the DB as a cache of the feature flag
 	// settings in case the management system is down when a pod starts.
-	if !s.persistFeatureFlags {
+
+	// Backing up feature flags section in case we need to restore them later on.
+	oldCfgFF := oldCfg.FeatureFlags
+	oldCfgNoEnvFF := oldCfgNoEnv.FeatureFlags
+	// Clearing FF sections to avoid both comparing and persisting them.
+	if s.readOnlyFF {
+		oldCfg.FeatureFlags = nil
 		newCfg.FeatureFlags = nil
+		newCfgNoEnv.FeatureFlags = nil
 	}
 
-	if err := s.backingStore.Set(newCfg); err != nil {
+	if err := s.backingStore.Set(newCfgNoEnv); err != nil {
 		return nil, errors.Wrap(err, "failed to persist")
 	}
 
-	if err := s.loadLockedWithOld(oldCfg, &unlockOnce); err != nil {
-		return nil, errors.Wrap(err, "failed to load on save")
+	newCfg = applyEnvironmentMap(newCfgNoEnv, GetEnvironment())
+	fixConfig(newCfg)
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	hasChanged, err := confsDiff(oldCfg, newCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compare configs")
+	}
+
+	// We restore the previously cleared feature flags sections back.
+	if s.readOnlyFF {
+		oldCfg.FeatureFlags = oldCfgFF
+		newCfg.FeatureFlags = oldCfgFF
+		newCfgNoEnv.FeatureFlags = oldCfgNoEnvFF
+	}
+
+	s.configNoEnv = newCfgNoEnv
+	s.config = newCfg
+
+	unlockOnce.Do(s.configLock.Unlock)
+
+	if hasChanged {
+		s.invokeConfigListeners(oldCfg, newCfg.Clone())
 	}
```

Comment:
Can we try to refactor this condition outside the `Set` and `Load` methods and remove the `unlockOnce` contraption? I believe it was added to work around the additional complexity of tracking different return paths, which is much simplified now. Would be good to get rid of this.

Reply:
I originally had the same sentiment although I can see how it's efficiently and cleanly used to avoid calling `configLock.Unlock()` every time we return early and at the same time allow for an unlocked code-path. Do you find that anti-idiomatic? To refactor this we would either need an additional wrapping method, some defer magic (it would probably get even more verbose) or just be okay with unlocking/locking before/after the listeners invocation.

Reply:
> Do you find that anti-idiomatic?

Yes, slightly. It is a trick which ultimately hurts readability in the long run IMO. I am okay with both wrapping or unlocking/locking.